rust/lance/src/dataset/write/retry.rs

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::future::Future;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use either::Either;
+use futures::future::FutureExt;
+use futures::TryFutureExt;
+use lance_core::utils::backoff::SlotBackoff;
+use lance_core::{Error, Result};
+use snafu::location;
+
+use crate::Dataset;
+
+/// Configuration for retry behavior
+#[derive(Debug, Clone)]
+pub struct RetryConfig {
+    pub max_retries: u32,
+    pub retry_timeout: Duration,
+}
+
+impl Default for RetryConfig {
+    fn default() -> Self {
+        Self {
+            max_retries: 10,
```

Comment:
we also set the defaults separately at delete, merge-insert, update level, and then there is another higher layer default in python. Since we have this shared retry config now, can we just make the upper levels configs optional so it all default to this automatically?

Reply:
We could. Honestly like having the defaults visible in the Python signature. That way users can see immediately what the default is, without having to look it up in the docs.